technologydata/currency_utils.py

```diff
@@ -0,0 +1,461 @@
+
+        Examples
+        --------
+        >>> CurrencyUtils.ensure_currency_unit("The price is USD-2025", r"[A-Z]{3}-\d{4}")
+        'USD-2025'
+        >>> CurrencyUtils.ensure_currency_unit("No currency here", r"[A-Z]{3}-\d{4}")
+        None
+
+        """
+        if not isinstance(input_string, str) or not isinstance(expected_format, str):
+            raise ValueError("Input must be a string.")
+
+        currency_unit_pattern = re.compile(expected_format)
+        match = currency_unit_pattern.search(input_string)
+        return match.group(0) if match else None
+
+    @staticmethod
+    def get_deflate_row_function(
+        base_year: int,
+        deflator_name: str,
+        target_currency: str,
+        year: str,
+        iso_code: str,
+        target_value_column: str,
+    ):
+        deflation_function = deflation_function_registry.get(deflator_name)
+        if deflation_function is None:
+            raise ValueError(
+                f"Deflator function '{deflator_name}' not found in registry"
+            )
+
+        def deflate_row(row):
+            row_df = pd.DataFrame([row])
+            deflated_df = deflation_function(
+                data=row_df,
+                base_year=base_year,
+                source_currency=row[iso_code],
+                target_currency=target_currency,
+                id_column=iso_code,
+                year_column=year,
+                value_column="value",
+                target_value_column=target_value_column,
+            )
+            return deflated_df.iloc[0]
+
+        return deflate_row
+
+    @staticmethod
+    def convert_and_adjust_currency(
+        base_year_val: int,
+        deflator_function_name: str,
+        target_currency_country_code: str,
+        pydeflate_path: pathlib.Path,
+        data: pd.DataFrame,
+    ) -> pd.DataFrame:
+        # Specify the path where deflator and exchange data will be saved
+        if pydeflate_path is not None:
+            pyd.set_pydeflate_path(pydeflate_path)
+        else:
+            raise ValueError(
+                "The path where the deflator and exchange data will be saved is None"
+            )
+
+        # Validate columns presence
+        required_columns = {"unit", "value", "region"}
+        if not required_columns.issubset(data.columns):
+            missing = required_columns - set(data.columns)
+            raise ValueError(f"Input DataFrame is missing required columns: {missing}")
+
+        # Create a copy of the original data to avoid modifying input directly
+        results = data.copy()
+
+        # Select rows that correspond to a unit column that fulfills the format <3-letter currency code>-<currency year>
+        has_currency_mask = (
+            results["unit"].apply(CurrencyUtils.ensure_currency_unit).notna()
+        )
+        currency_rows = results.loc[has_currency_mask].copy()
+
+        if currency_rows.empty:
+            raise ValueError("No rows contain a valid currency unit.")
```

Comment:
I think in that case it is fine to just `logger.warning`.
Then when we use the method in a context where unmodified rows should also be returned (cf. your decorator note), the application can still work as no rows would be modified and just a warning issues (rather than having to catch an ValueError with try-except